File: LAB 6/TheSutherland_HodgmanAlgorithm.py
import matplotlib.pyplot as plt
import numpy as np

XMIN, YMIN, XMAX, YMAX = 3, 3, 8, 8
POLYGON = np.array([[2, 2], [10, 2], [10, 10], [2, 10], [2, 2]])

def clip_polygon(POLY, X1, Y1, X2, Y2, EDGE):
    NEW_POLY = []
    N = len(POLY) - 1
    
    for I in range(N):
        CURRENT = POLY[I]
        NEXT = POLY[I + 1]
        
        CURRENT_INSIDE = False
        NEXT_INSIDE = False
        
        if EDGE == 'LEFT':
            CURRENT_INSIDE = CURRENT[0] >= XMIN
            NEXT_INSIDE = NEXT[0] >= XMIN
        elif EDGE == 'RIGHT':
            CURRENT_INSIDE = CURRENT[0] <= XMAX
            NEXT_INSIDE = NEXT[0] <= XMAX
        elif EDGE == 'BOTTOM':
            CURRENT_INSIDE = CURRENT[1] >= YMIN
            NEXT_INSIDE = NEXT[1] >= YMIN
        elif EDGE == 'TOP':
            CURRENT_INSIDE = CURRENT[1] <= YMAX
            NEXT_INSIDE = NEXT[1] <= YMAX
        
        if CURRENT_INSIDE and NEXT_INSIDE:
            NEW_POLY.append(NEXT)
        elif CURRENT_INSIDE and not NEXT_INSIDE:
            if EDGE == 'LEFT':
                T = (XMIN - CURRENT[0]) / (NEXT[0] - CURRENT[0])
                Y = CURRENT[1] + T * (NEXT[1] - CURRENT[1])
                NEW_POLY.append([XMIN, Y])
            elif EDGE == 'RIGHT':
                T = (XMAX - CURRENT[0]) / (NEXT[0] - CURRENT[0])
                Y = CURRENT[1] + T * (NEXT[1] - CURRENT[1])
                NEW_POLY.append([XMAX, Y])
            elif EDGE == 'BOTTOM':
                T = (YMIN - CURRENT[1]) / (NEXT[1] - CURRENT[1])
                X = CURRENT[0] + T * (NEXT[0] - CURRENT[0])
                NEW_POLY.append([X, YMIN])
            elif EDGE == 'TOP':
                T = (YMAX - CURRENT[1]) / (NEXT[1] - CURRENT[1])
                X = CURRENT[0] + T * (NEXT[0] - CURRENT[0])
                NEW_POLY.append([X, YMAX])
        elif not CURRENT_INSIDE and NEXT_INSIDE:
            if EDGE == 'LEFT':
                T = (XMIN - CURRENT[0]) / (NEXT[0] - CURRENT[0])
                Y = CURRENT[1] + T * (NEXT[1] - CURRENT[1])
                NEW_POLY.append([XMIN, Y])
            elif EDGE == 'RIGHT':
                T = (XMAX - CURRENT[0]) / (NEXT[0] - CURRENT[0])
                Y = CURRENT[1] + T * (NEXT[1] - CURRENT[1])
                NEW_POLY.append([XMAX, Y])
            elif EDGE == 'BOTTOM':
                T = (YMIN - CURRENT[1]) / (NEXT[1] - CURRENT[1])
                X = CURRENT[0] + T * (NEXT[0] - CURRENT[0])
                NEW_POLY.append([X, YMIN])
            elif EDGE == 'TOP':
                T = (YMAX - CURRENT[1]) / (NEXT[1] - CURRENT[1])
                X = CURRENT[0] + T * (NEXT[0] - CURRENT[0])
                NEW_POLY.append([X, YMAX])
            NEW_POLY.append(NEXT)
    
    if NEW_POLY:
        NEW_POLY.append(NEW_POLY[0])
    
    return np.array(NEW_POLY)

def sutherland_hodgman_clip(POLY):
    CLIPPED = POLY
    CLIPPED = clip_polygon(CLIPPED, XMIN, YMIN, XMAX, YMAX, 'LEFT')
    if len(CLIPPED) > 1:
        CLIPPED = clip_polygon(CLIPPED, XMIN, YMIN, XMAX, YMAX, 'RIGHT')
    if len(CLIPPED) > 1:
        CLIPPED = clip_polygon(CLIPPED, XMIN, YMIN, XMAX, YMAX, 'BOTTOM')
    if len(CLIPPED) > 1:
        CLIPPED = clip_polygon(CLIPPED, XMIN, YMIN, XMAX, YMAX, 'TOP')
    return CLIPPED

plt.figure(figsize=(10, 8))
plt.grid(True)
plt.title('Sutherland-Hodgman Polygon Clipping')
plt.xlabel('X-axis'); plt.ylabel('Y-axis')

plt.plot([XMIN, XMAX, XMAX, XMIN, XMIN], [YMIN, YMIN, YMAX, YMAX, YMIN], 'b-', linewidth=2, label='Clip Window')
plt.fill([XMIN, XMAX, XMAX, XMIN], [YMIN, YMIN, YMAX, YMAX], 'b', alpha=0.1)

plt.plot(POLYGON[:, 0], POLYGON[:, 1], 'r--', linewidth=2, label='Original Polygon')
plt.fill(POLYGON[:, 0], POLYGON[:, 1], 'r', alpha=0.2)

CLIPPED_POLY = sutherland_hodgman_clip(POLYGON)
if len(CLIPPED_POLY) > 1:
    plt.plot(CLIPPED_POLY[:, 0], CLIPPED_POLY[:, 1], 'g-', linewidth=2.5, label='Clipped Polygon')
    plt.fill(CLIPPED_POLY[:, 0], CLIPPED_POLY[:, 1], 'g', alpha=0.3)

plt.xlim(0, 12); plt.ylim(0, 12)
plt.legend()
plt.show()

File: LAB 6/The_Sutherland_Hodgman_Algorithm.m
clc; clear; close all;

XMIN = 3; YMIN = 3; XMAX = 8; YMAX = 8;
POLYGON = [2, 2; 10, 2; 10, 10; 2, 10; 2, 2];

function OUTPUT = clip_polygon_edge(INPUT, X1, Y1, X2, Y2, EDGE_TYPE, XMIN, YMIN, XMAX, YMAX)
    OUTPUT = [];
    N = size(INPUT, 1) - 1;
    
    for I = 1:N
        CURRENT = INPUT(I, :);
        NEXT = INPUT(I + 1, :);
        
        CURRENT_INSIDE = false;
        NEXT_INSIDE = false;
        
        switch EDGE_TYPE
            case 'LEFT'
                CURRENT_INSIDE = CURRENT(1) >= XMIN;
                NEXT_INSIDE = NEXT(1) >= XMIN;
            case 'RIGHT'
                CURRENT_INSIDE = CURRENT(1) <= XMAX;
                NEXT_INSIDE = NEXT(1) <= XMAX;
            case 'BOTTOM'
                CURRENT_INSIDE = CURRENT(2) >= YMIN;
                NEXT_INSIDE = NEXT(2) >= YMIN;
            case 'TOP'
                CURRENT_INSIDE = CURRENT(2) <= YMAX;
                NEXT_INSIDE = NEXT(2) <= YMAX;
        end
        
        if CURRENT_INSIDE && NEXT_INSIDE
            OUTPUT = [OUTPUT; NEXT];
        elseif CURRENT_INSIDE && ~NEXT_INSIDE
            switch EDGE_TYPE
                case 'LEFT'
                    T = (XMIN - CURRENT(1)) / (NEXT(1) - CURRENT(1));
                    Y = CURRENT(2) + T * (NEXT(2) - CURRENT(2));
                    OUTPUT = [OUTPUT; XMIN, Y];
                case 'RIGHT'
                    T = (XMAX - CURRENT(1)) / (NEXT(1) - CURRENT(1));
                    Y = CURRENT(2) + T * (NEXT(2) - CURRENT(2));
                    OUTPUT = [OUTPUT; XMAX, Y];
                case 'BOTTOM'
                    T = (YMIN - CURRENT(2)) / (NEXT(2) - CURRENT(2));
                    X = CURRENT(1) + T * (NEXT(1) - CURRENT(1));
                    OUTPUT = [OUTPUT; X, YMIN];
                case 'TOP'
                    T = (YMAX - CURRENT(2)) / (NEXT(2) - CURRENT(2));
                    X = CURRENT(1) + T * (NEXT(1) - CURRENT(1));
                    OUTPUT = [OUTPUT; X, YMAX];
            end
        elseif ~CURRENT_INSIDE && NEXT_INSIDE
            switch EDGE_TYPE
                case 'LEFT'
                    T = (XMIN - CURRENT(1)) / (NEXT(1) - CURRENT(1));
                    Y = CURRENT(2) + T * (NEXT(2) - CURRENT(2));
                    OUTPUT = [OUTPUT; XMIN, Y];
                case 'RIGHT'
                    T = (XMAX - CURRENT(1)) / (NEXT(1) - CURRENT(1));
                    Y = CURRENT(2) + T * (NEXT(2) - CURRENT(2));
                    OUTPUT = [OUTPUT; XMAX, Y];
                case 'BOTTOM'
                    T = (YMIN - CURRENT(2)) / (NEXT(2) - CURRENT(2));
                    X = CURRENT(1) + T * (NEXT(1) - CURRENT(1));
                    OUTPUT = [OUTPUT; X, YMIN];
                case 'TOP'
                    T = (YMAX - CURRENT(2)) / (NEXT(2) - CURRENT(2));
                    X = CURRENT(1) + T * (NEXT(1) - CURRENT(1));
                    OUTPUT = [OUTPUT; X, YMAX];
            end
            OUTPUT = [OUTPUT; NEXT];
        end
    end
    
    if ~isempty(OUTPUT)
        OUTPUT = [OUTPUT; OUTPUT(1, :)];
    end
end

function CLIPPED = sutherland_hodgman_clip(POLY, XMIN, YMIN, XMAX, YMAX)
    CLIPPED = POLY;
    
    CLIPPED = clip_polygon_edge(CLIPPED, XMIN, YMIN, XMAX, YMAX, 'LEFT', XMIN, YMIN, XMAX, YMAX);
    if size(CLIPPED, 1) <= 1
        return;
    end
    
    CLIPPED = clip_polygon_edge(CLIPPED, XMIN, YMIN, XMAX, YMAX, 'RIGHT', XMIN, YMIN, XMAX, YMAX);
    if size(CLIPPED, 1) <= 1
        return;
    end
    
    CLIPPED = clip_polygon_edge(CLIPPED, XMIN, YMIN, XMAX, YMAX, 'BOTTOM', XMIN, YMIN, XMAX, YMAX);
    if size(CLIPPED, 1) <= 1
        return;
    end
    
    CLIPPED = clip_polygon_edge(CLIPPED, XMIN, YMIN, XMAX, YMAX, 'TOP', XMIN, YMIN, XMAX, YMAX);
end

figure('Position', [100, 100, 1000, 400]);

subplot(1, 2, 1);
hold on; grid on; axis equal;
title('Sutherland-Hodgman: Before Clipping');
xlabel('X-axis'); ylabel('Y-axis');

% Draw clipping window
plot([XMIN, XMAX, XMAX, XMIN, XMIN], [YMIN, YMIN, YMAX, YMAX, YMIN], 'b-', 'LineWidth', 2, 'DisplayName', 'Clip Window');
fill([XMIN, XMAX, XMAX, XMIN], [YMIN, YMIN, YMAX, YMAX], 'b', 'FaceAlpha', 0.1);

% Draw original polygon
plot(POLYGON(:, 1), POLYGON(:, 2), 'r-', 'LineWidth', 2, 'DisplayName', 'Original Polygon');
fill(POLYGON(:, 1), POLYGON(:, 2), 'r', 'FaceAlpha', 0.3);

xlim([0, 12]); ylim([0, 12]);
legend('Location', 'best');

subplot(1, 2, 2);
hold on; grid on; axis equal;
title('Sutherland-Hodgman: After Clipping');
xlabel('X-axis'); ylabel('Y-axis');

% Draw clipping window
plot([XMIN, XMAX, XMAX, XMIN, XMIN], [YMIN, YMIN, YMAX, YMAX, YMIN], 'b-', 'LineWidth', 2, 'DisplayName', 'Clip Window');
fill([XMIN, XMAX, XMAX, XMIN], [YMIN, YMIN, YMAX, YMAX], 'b', 'FaceAlpha', 0.1);

% Clip and draw
CLIPPED_POLY = sutherland_hodgman_clip(POLYGON, XMIN, YMIN, XMAX, YMAX);
if size(CLIPPED_POLY, 1) > 1
    plot(CLIPPED_POLY(:, 1), CLIPPED_POLY(:, 2), 'g-', 'LineWidth', 2.5, 'DisplayName', 'Clipped Polygon');
    fill(CLIPPED_POLY(:, 1), CLIPPED_POLY(:, 2), 'g', 'FaceAlpha', 0.5);
end

xlim([0, 12]); ylim([0, 12]);
legend('Location', 'best');
